include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,707 @@
+#else
+#if __cplusplus >= 201703
+#include <execution>
+#endif
+#endif
+
+namespace boost { namespace math { namespace detail
+{
+// https://mathworld.wolfram.com/SieveofEratosthenes.html
+// https://www.cs.utexas.edu/users/misra/scannedPdf.dir/linearSieve.pdf
+template<class Z, class Container>
+void linear_sieve(Z upper_bound, Container &c)
+{
+    size_t least_divisors_size{static_cast<size_t>(upper_bound + 1)};
+    std::unique_ptr<Z[]> least_divisors{new Z[least_divisors_size]{0}};
+
+    for (Z i{2}; i <= upper_bound; ++i)
+    {
+        if (least_divisors[i] == 0)
+        {
+            least_divisors[i] = i;
+            c.emplace_back(i);
+        }
+
+        for (size_t j{}; j < least_divisors_size; ++j)
+        {
+            if (j >= c.size())
+            {
+                break;
+            }
+
+            else if (c[j] > least_divisors[i])
+            {
+                break;
+            }
+
+            else if (i * c[j] > upper_bound)
+            {
+                break;
+            }
+
+            else
+            {
+                least_divisors[i * c[j]] = c[j];
+            }
+        }
+    }
+}
+
+//https://core.ac.uk/download/pdf/62440589.pdf
+template<class Z, class PrimeContainer, class Container>
+void mask_sieve(Z lower_bound, Z upper_bound, const PrimeContainer& primes, Container &c)
+{
+    Z limit {static_cast<Z>(std::floor(std::sqrt(static_cast<double>(upper_bound)))) + 1};
+
+    size_t primes_size {};
+    auto it{primes.begin()};
+    while(it != primes.end() && *it < limit)
+    {
+        ++primes_size;
+        ++it;
+    }
+
+    const size_t n {static_cast<size_t>(upper_bound - lower_bound + 1)};
+    std::unique_ptr<bool[]> is_prime {new bool[n]};
+    memset(is_prime.get(), true, sizeof(*is_prime.get()) * (n));
+
+    #if __cplusplus >= 201703 || _MSVC_LANG >= 201703
+    // Enable use of thread pool, not SIMD compatible
+    std::for_each(std::execution::par, primes.begin(), it, [&is_prime, lower_bound, upper_bound](auto prime)
+    {
+        for(Z j {std::max(prime * prime, (lower_bound + prime - 1) / prime * prime)}; j <= upper_bound; j += prime)
+        {
+            is_prime[j - lower_bound] = false;
+        }
+    });
+    #else
+
+    for(size_t i{}; i < primes_size; ++i)
+    {
+        Z current_prime{primes[i]};
+        for(Z j {std::max(current_prime * current_prime, (lower_bound + current_prime - 1) / current_prime * current_prime)};
+            j <= upper_bound; j += current_prime)
+        {
+            is_prime[j - lower_bound] = false;
+        }
+    }
+    #endif
+
+
+    if(lower_bound == 1)
+    {
+        is_prime[0] = false;
+    }
+
+    for(Z i{lower_bound}; i <= upper_bound; ++i)
+    {
+        if(is_prime[i - lower_bound])
+        {
+            c.emplace_back(i);
+        }
+    }
+}
+
+// https://minds.wisconsin.edu/bitstream/handle/1793/59248/TR909.pdf?sequence=1
+// Implementing S
+template<class Z>
+class SetS
+{
+private:
+    std::deque<Z> srec_;
```

Comment:
Have you benchmarked the benefit of using `std::deque` as opposed to `std::vector`?

Reply:
I had not because the performance was so far off of what I expected. There is a lot of searching so it would not surprise me if the `std::vector` is faster.

Reply:
This guy did a great benchmark comparison years ago: https://baptiste-wicht.com/posts/2012/12/cpp-benchmark-vector-list-deque.html

OK, wow, that was 8 years ago already.